mgr: add allowlist to get_all_perf_counters()

epuertat opened the pull request:
## TODO

- [x] Add `allowlist` parameter to mgr API get_all_perf_counters().
- [x] Provide allowlisted counters for Prometheus exporter based on existing Grafana and Alertmanage queries.
- [x] Add gzip support to the /metrics endpoint (it reduces 98% the size of  the HTTP response, at the expense of halving the max request rate, but that doesn't seem to be the bottleneck here).
- [ ] Review non-perf-counter metrics.
- [ ] Expose metrics in different endpoints (e.g.: `/perf-counters`, `/health`). This would provide horizontal scalability and the scrape interval could be customized based on the data set size vs. data freshness. This could also enable the ["split by use" scaling in Prometheus](https://www.robustperception.io/scaling-and-federating-prometheus).


## Results

For a vstart cluster with 3-mon, 3-mgr, 60-osd, 1 rgw daemon, after this PR:
* **x8 sample reduction**: from 14,312 to 1,773 metrics/scrape,
* **x8.7 payload reduction**: from 978,849 down to 113,537 bytes/scrape,
* **x88 compressed payload reduction**: 11,116 byes/scrape,
* **x11.7 mgr API calls reduction**: from 10,776 to 916 calls/scrape.

## Caveats

From the list of metrics provided by @pereman2, in this environment (vstart-based) I couldn't get the following ones:
```
ceph_daemon not found
ceph_data_sync_from_zone_fetch_bytes_sum not found
ceph_data_sync_from_zone_fetch_bytes_count not found
ceph_data_sync_from_zone_poll_latency_sum not found
ceph_data_sync_from_zone_fetch_errors not found
ceph_rbd_write_ops not found
ceph_rbd_read_ops not found
ceph_rbd_write_bytes not found
ceph_rbd_read_bytes not found
ceph_rbd_write_latency_sum not found
ceph_rbd_write_latency_count not found
ceph_rbd_read_latency_sum not found
ceph_rbd_read_latency_count not found
ceph_pg_undersized not found
ceph_pg_degraded not found
ceph_pg_inconsistent not found
ceph_pg_down not found
ceph_hosts not found
ceph_disk_occupation not found
ceph_disk_occupation_human not found
ceph_disk_latency_read not found
ceph_node_disk_rate not found
ceph_disk_latency_write not found
ceph_bluestore_bluestore_compressed_original not found
ceph_bluestore_bluestore_compressed_allocated not found
ceph_node_down not found
ceph_version not found
ceph_disk_latency not found
ceph_node_disk_irate not found
ceph_pools_iops not found
ceph_pools_iops_bytes not found
ceph_versions_running not found
```

Fixes: https://tracker.ceph.com/issues/54403
Signed-off-by: Ernesto Puerta <epuertat@redhat.com>





<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".

  - The Signed-off-by line in every git commit is important; see <span class="x x-first x-last">[Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/master/</span>SubmittingPatches.rst<span class="x x-first x-last">)</span>
-->

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

pereman2 reviewed (commented):
Those are :1st_place_medal: numbers.

I think that if this allow list were to be placed in the c++ the payload reduction would be even more effective since there would be less data to serialize and therefore, reduce the impact of calling get perf counters. 

If I remember correctly, in Paul's findings we saw that some of the calls like `get_fs`, without FSs at all, would take more time than those calls that actually should take more time because of locking issues or subinterpreters in general.

This is just a quick assumption so it should be taken with a grain of salt.

avanthakkar commented:
> Those are 1st_place_medal numbers.
> 
> I think that if this allow list were to be placed in the c++ the payload reduction would be even more effective since there would be less data to serialize and therefore, reduce the impact of calling get perf counters.
Yes, but the mgr is currently keeping track of all the updated counters(with `mgr_stats_period`), right? But yeah prometheus may not be using all of them, so I'm not sure how we can specify allow list at c++ level currently. 
> 
> If I remember correctly, in Paul's findings we saw that some of the calls like `get_fs`, without FSs at all, would take more time than those calls that actually should take more time because of locking issues or subinterpreters in general.
> 
> This is just a quick assumption so it should be taken with a grain of salt.

avanthakkar commented on src/pybind/mgr/prometheus/module.py:
```diff
@@ -53,7 +53,8 @@ def os_exit_noop(status: int) -> None:
 cherrypy.config.update({
-    'response.headers.server': 'Ceph-Prometheus'
+    'response.headers.server': 'Ceph-Prometheus',
+    'tools.gzip.on': True,
```
@epuertat  Maybe adding `cache-control` too? Not sure if it'll impact performance much or not.

epuertat commented:
> Those are 1st_place_medal numbers.

Yup, hopefully this allows to scale to the 8k mark with no noticeable performance impact.

> I think that if this allow list were to be placed in the c++ the payload reduction would be even more effective since there would be less data to serialize and therefore, reduce the impact of calling get perf counters.

I thought the same initially, because I assumed that the `get_all_perf_counters()` mgr call was internally a C++ one, but the truth is that this call is Python only: it actually uses the most fine-grained API call `get_counter()`. So in this very case, perhaps the issue is caused the the high overhead of such a fine-grained call. For 60 OSDs this results in 10,776 `get_latest_counter()` calls, which translates into 180k calls for a 1k OSD cluster. That's a lot. After this PR, that would be just 15,200 calls, slightly above the current mark.

This is the pseudo-code for the `get_all_perf_counters()`:
```
  for srv in mgr.list_servers:
    for srvc in srv.services:
      for counter_path, counter_schema in mgr.get_perf_schema(srvc.type, srvc.id):
        if counter_schema.prio < prio_limit: skip
        if counter_schema.tags != tags: skip
        
        if LONGRUNGAVG:
          counter_info[value], counter_info[count] = mgr.get_latest_avg(srvc.type, srvc.id, counter_path)
          result[srvc.type + srvc.id][counter_path] = counter_info
        else:
          counter_info[value] = mgr.get_latest(srvc.type, srvc.id, counter_path)
          result[srvc.type + srvc.id][counter_path] = counter_info
  return result
```

This could be improved, paradoxically, by providing a more coarse-grained call, perhaps: `get_service_latest_counters()` or `get_host_latest_counters()` to retrieve all the counters for a given service or host.

> If I remember correctly, in Paul's findings we saw that some of the calls like `get_fs`, without FSs at all, would take more time than those calls that actually should take more time because of locking issues or subinterpreters in general.

Thanks! I have to take a look at that, since I only focused here on perf-counters.

epuertat commented:
> > Those are 1st_place_medal numbers.
> > > I think that if this allow list were to be placed in the c++ the payload reduction would be even more effective since there would be less data to serialize and therefore, reduce the impact of calling get perf counters.
> 
> > If I remember correctly, in Paul's findings we saw that some of the calls like `get_fs`, without FSs at all, would take more time than those calls that actually should take more time because of locking issues or subinterpreters in general.
> > This is just a quick assumption so it should be taken with a grain of salt.
> 
> Yes, but the mgr is currently keeping track of all the updated counters(with `mgr_stats_period`), right? But yeah prometheus may not be using all of them, so I'm not sure how we can specify allow list at c++ level currently.

I started this PR https://github.com/ceph/ceph/pull/45150, but then I checked the code and saw that this could be done in Python, as the perf-counters are not serialized at once, but cherry-picked.

Different issue is whether to reduce the amount of metrics sent from the daemons to the mgr, as that could also contribute to overloading the mgr or causing bursts in the network.

pcuzner commented:
+1 for the gzip usage! What does that do for cpu?

I mentioned using an allowlist before. In my tests I could scale to ~ 8000 OSDs with consistent scrapes (without gzip), but beyond 8000 I started seeing 6-8 failures every hour and spikes in mgr/prometheus data gathering to over 4s - since the data sent to the mgr itself remains unchanged.

Just not sure this addresses the issues we have, and as @epuertat pointed out elsewhere we also have to be mindful of influx and telegraf too.

epuertat commented:
> +1 for the gzip usage! What does that do for cpu?

I ran an Apache benchmark with compression (`ab -c 4 -klt 20 -H "Accept-Encoding: gzip,deflate"  http://<host>:9283/metrics`) and I got 418 reqs/s with a ~70% CPU. Without gzip the same benchmark gives 550 reqs/s and ~55% CPU.

With the current optimizations, the 8k OSD would generate ~188 MB of raw data, and the compression speed (gzip level=5) in my 4-year old laptop it takes 6 secs to compress 180 MB of random data (`head -c 180M </dev/urandom >random && time gzip -c random > /dev/null`). I assume this would be much faster on a brand-new server.

In any case, Cherrypy supports compressing streaming responses (which would speed up the response and also save a lot of memory).

Additionally, splitting the single `/metrics` endpoint in multiple ones (I'm exploring that now) could work instead. It's supported with [the `__metrics_path__` label](https://github.com/prometheus/prometheus/issues/4001#issuecomment-375426454).

> I mentioned using an allowlist before. In my tests I could scale to ~ 8000 OSDs with consistent scrapes (without gzip), but beyond 8000 I started seeing 6-8 failures every hour and spikes in mgr/prometheus data gathering to over 4s - since the data sent to the mgr itself remains unchanged.

Yes, I assumed you used a similar approach as this one.

> Just not sure this addresses the issues we have, and as @epuertat pointed out elsewhere we also have to be mindful of influx and telegraf too.

I don't think Telegraf or Influx will face the exact same issue since they both push metrics to the metric collector, and use a less verbose message format compared to Prometheus.
